### PR TITLE
intial commit for ceph-cosi-driver

### PR DESCRIPTION
**Pull request by thotz**
This is the first step towards COSI driver implementation for ceph.
Signed-off-by: Jiffin Tony Thottan <thottanjiffin@gmail.com>

### CODE REVIEW

**Comment by thotz** [comment]
Currently creation of BAR are not working atm, https://github.com/kubernetes-sigs/container-object-storage-interface-controller/pull/55

File: LICENSE [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+      and charge a fee for, acceptance of support, warranty, indemnity,
+      or other liability obligations and/or rights consistent with this
+      License. However, in accepting such obligations, You may act only
+      on Your own behalf and on Your sole responsibility, not on behalf
+      of any other Contributor, and only if You agree to indemnify,
+      defend, and hold each Contributor harmless for any liability
+      incurred by, or claims asserted against, such Contributor by reason
+      of your accepting any such warranty or additional liability.
+
+   END OF TERMS AND CONDITIONS
```
Comment (leseb) [inline LICENSE:176]: Why is the `APPENDIX` missing?

File: cmd/ceph-cosi-driver/cmd.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
+// Copyright 2021 The Ceph-COSI Authors.
```
Comment (leseb) [inline cmd/ceph-cosi-driver/cmd.go:1]: Use `/*`instead?

File: pkg/util/s3client/s3-handlers.go [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+			switch aerr.Code() {
+			case s3.ErrCodeBucketAlreadyExists:
+				klog.Infof("bucket %q already exists", name)
+				return nil
+			case s3.ErrCodeBucketAlreadyOwnedByYou:
+				klog.Infof("bucket %q already owned by you", name)
+				return nil
+			}
+		}
+		return errors.Wrapf(err, "failed to create bucket %q", name)
```
Comment (leseb) [inline pkg/util/s3client/s3-handlers.go:97]: I know Rook is using this `"github.com/pkg/errors"` package but to wrap an error you can just use the `fmt` package with the `%w` keyword instead.

File: .gitignore [hunk 4]
```diff
@@ -1 +1,8 @@
-
+*.tmp
+.DS_Store
+.build
+*.swp
+travis.yml
```
Comment (leseb) [inline .gitignore:5]: Why not using github actions?

File: .gitignore [hunk 5]
```diff
@@ -1 +1,8 @@
...
+bin
+.idea
+vendor
```
Comment (leseb) [inline .gitignore:8]: Are we vendoring?

File: go.mod [hunk 6]
```diff
@@ -0,0 +1,18 @@
+module github.com/thotz/cosi-driver-ceph
+
+go 1.15
+
+require (
+	github.com/aws/aws-sdk-go v1.38.24
+	github.com/ceph/go-ceph v0.9.1-0.20210531154531-6be8d370cbd0
+	github.com/google/uuid v1.2.0 // indirect
+	github.com/pkg/errors v0.9.1
```
Comment (leseb) [inline go.mod:9]: based on my comment below I'd drop this package and just use `errors`.

**Review by leseb — CHANGES_REQUESTED** [review]
Briefly review, thanks for all the work!

File: cmd/ceph-cosi-driver/main.go [hunk 7]
```diff
@@ -0,0 +1,47 @@
+/*
+Copyright 2021 The Ceph-COSI Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
```
[the remark this answers is not in the document]
Comment (leseb) [reply cmd/ceph-cosi-driver/main.go:3]: See cmd/ceph-cosi-driver/cmd.go

File: pkg/driver/driver.go [hunk 8]
```diff
@@ -0,0 +1,44 @@
+/*
+Copyright 2021 The Ceph-COSI Authors.
```
Comment (leseb) [inline pkg/driver/driver.go:2]: ditto see cmd/ceph-cosi-driver/cmd.go

File: pkg/driver/driver.go [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package driver
+
+import (
+	"context"
+
+	radosgwapi "github.com/ceph/go-ceph/rgw/admin"
```
Comment (leseb) [inline pkg/driver/driver.go:21]:
```suggestion
	rgwadmin "github.com/ceph/go-ceph/rgw/admin"
```

File: pkg/driver/driver.go [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+	"github.com/thotz/cosi-driver-ceph/pkg/util/s3client"
```
Comment (leseb) [inline pkg/driver/driver.go:22]: What is this repo? Why is it not in the ceph org?

File: pkg/driver/driver.go [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+	"k8s.io/klog/v2"
+)
+
+func NewDriver(ctx context.Context, provisioner, rgwEndpoint, accessKey, secretKey string) (*IdentityServer, *ProvisionerServer, error) {
+	// TODO : use different user this operation
+	s3Client, err := s3client.NewS3Agent(accessKey, secretKey, rgwEndpoint, true)
+	if err != nil {
+		klog.Fatalln(err)
+	}
+	//TODO : add support for TLS endpoint
```
Comment (leseb) [inline pkg/driver/driver.go:32]: +1

File: pkg/driver/identityserver.go [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+type IdentityServer struct {
+	provisioner string
+}
+
+func (id *IdentityServer) ProvisionerGetInfo(ctx context.Context,
+	req *cosi.ProvisionerGetInfoRequest) (*cosi.ProvisionerGetInfoResponse, error) {
+
+	if id.provisioner == "" {
+		klog.ErrorS(fmt.Errorf("provisioner name cannot be empty"), "Invalid argument")
+		return nil, status.Error(codes.InvalidArgument, "ProvisionerName is empty")
```
Comment (leseb) [inline pkg/driver/identityserver.go:38]:
```suggestion
		return nil, status.Error(codes.InvalidArgument, "Provisioner name is empty")
```

File: pkg/driver/identityserver.go [hunk 13]
```diff
@@ -0,0 +1,44 @@
+/*
+Copyright 2021 The Ceph-COSI Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (leseb) [inline pkg/driver/identityserver.go:3]: ditto see cmd/ceph-cosi-driver/cmd.go

File: pkg/driver/provisioner.go [hunk 14]
```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2021 The Ceph-COSI Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (leseb) [inline pkg/driver/provisioner.go:3]: ditto see cmd/ceph-cosi-driver/cmd.go

File: pkg/driver/provisioner.go [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+*/
+
+package driver
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	radosgwapi "github.com/ceph/go-ceph/rgw/admin"
```
Comment (leseb) [inline pkg/driver/provisioner.go:23]:
```suggestion
	rgwadmin "github.com/ceph/go-ceph/rgw/admin"
```

File: pkg/driver/provisioner.go [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+	cosi "sigs.k8s.io/container-object-storage-interface-spec"
+)
+
+// contains two clients
+// 1.) for RGWAdminOps : mainly for user related operations
+// 2.) for S3 operations : mainly for bucket related operations
+type ProvisionerServer struct {
+	provisioner        string
+	S3Client           *s3client.S3Agent
+	radosgwAdminClient *radosgwapi.API
```
Comment (leseb) [inline pkg/driver/provisioner.go:37]:
```suggestion
	rgwAdminClient *radosgwapi.API
```

File: pkg/driver/provisioner.go [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+	"k8s.io/klog/v2"
+	cosi "sigs.k8s.io/container-object-storage-interface-spec"
+)
+
+// contains two clients
+// 1.) for RGWAdminOps : mainly for user related operations
+// 2.) for S3 operations : mainly for bucket related operations
+type ProvisionerServer struct {
+	provisioner        string
+	S3Client           *s3client.S3Agent
```
Comment (leseb) [inline pkg/driver/provisioner.go:36]: Does it need to be exported?

File: pkg/driver/provisioner.go [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+// ProvisionerCreateBucket is an idempotent method for creating buckets
+// It is expected to create the same bucket given a bucketName and protocol
+// If the bucket already exists, then it MUST return codes.AlreadyExists
+// Return values
+//    nil -                   Bucket successfully created
+//    codes.AlreadyExists -   Bucket already exists. No more retries
+//    non-nil err -           Internal error                                [requeue'd with exponential backoff]
+func (s *ProvisionerServer) ProvisionerCreateBucket(ctx context.Context,
+	req *cosi.ProvisionerCreateBucketRequest) (*cosi.ProvisionerCreateBucketResponse, error) {
+	klog.Info("Using ceph rgw to create Backend Bucket")
```
Comment (leseb) [inline pkg/driver/provisioner.go:49]:
```suggestion
	klog.Info("Using Ceph RADOS Gateway to create Backend Bucket")
```

File: pkg/driver/provisioner.go [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+	protocol := req.GetProtocol()
+	if protocol == nil {
+		klog.ErrorS(fmt.Errorf("Invalid Argument"), "Protocol is nil")
```
Comment (leseb) [inline pkg/driver/provisioner.go:52]: Is `Invalid Argument` the most appropriate error?

Comment (leseb) [reply pkg/driver/provisioner.go:52]: Also please register these errors as variables like:

```go
var (
  errNilProtocol = errors.New("protocol is nil")
)

File: pkg/driver/provisioner.go [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+		return nil, status.Error(codes.InvalidArgument, "Protocol is nil")
+	}
+	s3 := protocol.GetS3()
+	if s3 == nil {
+		klog.ErrorS(fmt.Errorf("Invalid Argument"), "S3 protocol is nil")
```
Comment (leseb) [inline pkg/driver/provisioner.go:57]: ditto use register the error in an errors.go file?

File: pkg/driver/provisioner.go [hunk 21]
```diff
@@ -0,0 +1,151 @@
...
+		return nil, status.Error(codes.InvalidArgument, "S3 Protocol is nil")
+	}
+	//TODO : validate S3 protocol defined, check points valid rgwendpoint, v4 signature check etc
+	bucketName := req.GetName()
+	klog.V(3).InfoS("Create Bucket", "name", bucketName)
```
Comment (leseb) [inline pkg/driver/provisioner.go:62]:
```suggestion
	klog.V(3).InfoS("Creating Bucket", "name", bucketName)
```

File: pkg/driver/provisioner.go [hunk 22]
```diff
@@ -0,0 +1,151 @@
...
+
+	err := s.S3Client.CreateBucket(bucketName)
+	if err != nil {
+		// Check to see if the bucket already exists by above api
+		klog.ErrorS(err, "Bucket creation failed")
```
Comment (leseb) [inline pkg/driver/provisioner.go:67]:
```suggestion
		klog.ErrorS(err, "failed to create bucket")
```

Also pass the `bucketName`?

File: pkg/driver/provisioner.go [hunk 23]
```diff
@@ -0,0 +1,151 @@
...
+	klog.Infof("Successfully created Backend Bucket %q", bucketName)
+
+	return &cosi.ProvisionerCreateBucketResponse{
+		BucketId: bucketName,
+	}, nil
+}
+
+func (s *ProvisionerServer) ProvisionerDeleteBucket(ctx context.Context,
+	req *cosi.ProvisionerDeleteBucketRequest) (*cosi.ProvisionerDeleteBucketResponse, error) {
+	klog.Infof("Delete bucket %q", req.GetBucketId())
```
Comment (leseb) [inline pkg/driver/provisioner.go:79]:
```suggestion
	klog.Infof("Deleting bucket %q", req.GetBucketId())
```

File: pkg/driver/provisioner.go [hunk 24]
```diff
@@ -0,0 +1,151 @@
...
+	if _, err := s.S3Client.DeleteBucket(req.GetBucketId()); err != nil {
+		klog.Info("failed to delete bucket %q", req.GetBucketId())
```
Comment (leseb) [inline pkg/driver/provisioner.go:81]:
```suggestion
		klog.ErrorS("failed to delete bucket %q", req.GetBucketId())
```

File: pkg/driver/provisioner.go [hunk 25]
```diff
@@ -0,0 +1,151 @@
...
+		return nil, status.Error(codes.Internal, "Bucket deletion failed")
```
Comment (leseb) [inline pkg/driver/provisioner.go:82]: `failed to`

File: pkg/driver/provisioner.go [hunk 26]
```diff
@@ -0,0 +1,151 @@
...
+	return &cosi.ProvisionerDeleteBucketResponse{}, nil
+}
+
+func (s *ProvisionerServer) ProvisionerGrantBucketAccess(ctx context.Context,
+	req *cosi.ProvisionerGrantBucketAccessRequest) (*cosi.ProvisionerGrantBucketAccessResponse, error) {
+	userName := req.GetAccountName()
+	bucketName := req.GetBucketId()
+	accessPolicy := req.GetAccessPolicy()
+	klog.Info("Granting user %q the policy %q to bucket %q", userName, bucketName, accessPolicy)
+	user, err := s.radosgwAdminClient.CreateUser(context.Background(), radosgwapi.User{
```
Comment (leseb) [inline pkg/driver/provisioner.go:94]: Why not passing `ctx` instead of a new one?

File: pkg/driver/provisioner.go [hunk 27]
```diff
@@ -0,0 +1,151 @@
...
+		ID:          userName,
+		DisplayName: userName,
+	})
+	if err != nil {
+		klog.Error("failed to create user", err)
```
Comment (leseb) [inline pkg/driver/provisioner.go:99]: Should we check for `UserAlreadyExists`?

File: pkg/driver/provisioner.go [hunk 28]
```diff
@@ -0,0 +1,151 @@
...
+		policy = s3client.NewBucketPolicy(*statement)
+	} else {
+		policy = policy.ModifyBucketPolicy(*statement)
+	}
+	out, err := s.S3Client.PutBucketPolicy(bucketName, *policy)
+	if err != nil {
+		klog.Error("failed to set policy", err)
+		return nil, status.Error(codes.Internal, "puting policy failed")
+	}
+	klog.Infof("set policy %v", out)
```
Comment (leseb) [inline pkg/driver/provisioner.go:127]: is it intended?

File: pkg/driver/provisioner.go [hunk 29]
```diff
@@ -0,0 +1,151 @@
...
+
+	// TODO : limit the bucket count for this user to 0
+
+	// Below response if not final, may change in future
+	return &cosi.ProvisionerGrantBucketAccessResponse{
+		AccountId:   userName,
+		Credentials: fmt.Sprintf("[default]\naws_access_key %s\naws_secret_key %s", user.Keys[0].AccessKey, user.Keys[0].SecretKey),
```
Comment (leseb) [inline pkg/driver/provisioner.go:134]: A function will do nice instead. To return the credentials

File: pkg/driver/provisioner.go [hunk 30]
```diff
@@ -0,0 +1,151 @@
...
+	}, nil
+}
+
+func (s *ProvisionerServer) ProvisionerRevokeBucketAccess(ctx context.Context,
+	req *cosi.ProvisionerRevokeBucketAccessRequest) (*cosi.ProvisionerRevokeBucketAccessResponse, error) {
+
+	// TODO : instead of deleting user, revoke its permission and delete only if no more bucket attached to it
+	klog.Infof("Delete user %q", req.GetAccountId())
```
Comment (leseb) [inline pkg/driver/provisioner.go:142]:
```suggestion
	klog.Infof("Deleting user %q", req.GetAccountId())
```

File: pkg/driver/provisioner.go [hunk 31]
```diff
@@ -0,0 +1,151 @@
...
+	if err := s.radosgwAdminClient.RemoveUser(context.Background(), radosgwapi.User{
+		ID:          req.GetAccountId(),
+		DisplayName: req.GetAccountId(),
+	}); err != nil {
+		klog.Error("falied to Revoke Bucket Access")
```
Comment (leseb) [inline pkg/driver/provisioner.go:147]:
```suggestion
		klog.Error("failed to Revoke Bucket Access")
```

File: pkg/util/s3client/policy.go [hunk 32]
```diff
@@ -0,0 +1,323 @@
+/*
+Copyright 2021 The Ceph-COSI Authors.
```
Comment (leseb) [inline pkg/util/s3client/policy.go:2]: ditto see cmd/ceph-cosi-driver/cmd.go

File: pkg/util/s3client/policy.go [hunk 33]
```diff
@@ -0,0 +1,323 @@
...
+		for j, v := range principals {
+			if u == v {
+				principals = append(principals[:j], principals[:j+1]...)
+			}
+		}
+	}
+	ps.Principal[awsPrinciple] = principals
+}
+
+// //////////////
```
Comment (leseb) [inline pkg/util/s3client/policy.go:321]: let's remove this

File: pkg/util/s3client/s3-handlers.go [hunk 34]
```diff
@@ -0,0 +1,181 @@
+/*
+Copyright 2021 The Ceph-COSI Authors.
```
Comment (leseb) [inline pkg/util/s3client/s3-handlers.go:2]: ditto see cmd/ceph-cosi-driver/cmd.go

**Review by leseb — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md [hunk 35]
```diff
@@ -0,0 +1,26 @@
+# Contributing Guidelines
+
+Welcome to Kubernetes. We are excited about the prospect of you joining our [community](https://git.k8s.io/community)! The Kubernetes community abides by the CNCF [code of conduct](code-of-conduct.md). Here is an excerpt:
```
[the remark this answers is not in the document]
Comment (thotz) [reply CONTRIBUTING.md]: removed this file for the time being

File: cmd/ceph-cosi-driver/cmd.go [hunk 36]
```diff
@@ -0,0 +1,116 @@
...
+		Endpoint,
+		"endpoint where rgw server is listening")
+
+	stringFlag(&AccessKey,
+		"accesskey",
+		"a",
+		AccessKey,
+		"access key for rgw")
+
+	stringFlag(&SecretKey,
```
[the remark this answers is not in the document]
Comment (thotz) [reply cmd/ceph-cosi-driver/cmd.go:84]: Yaa we need options for that, will address later

File: pkg/driver/provisioner.go (hunk 17 above)
Comment (thotz) [reply pkg/driver/provisioner.go]: Nope

File: pkg/driver/provisioner.go [hunk 37]
```diff
@@ -0,0 +1,158 @@
...
+}
+
+func (s *ProvisionerServer) ProvisionerDeleteBucket(ctx context.Context,
+	req *cosi.ProvisionerDeleteBucketRequest) (*cosi.ProvisionerDeleteBucketResponse, error) {
+	klog.Infof("Deleting bucket %q", req.GetBucketId())
+	if _, err := s.s3Client.DeleteBucket(req.GetBucketId()); err != nil {
+		klog.ErrorS(err, "failed to delete bucket %q", req.GetBucketId())
+		return nil, status.Error(codes.Internal, "failed to delete bucket")
+	}
+
```
Comment (leseb) [inline pkg/driver/provisioner.go:85]: Let's log successful deletion.

File: pkg/driver/provisioner.go [hunk 38]
```diff
@@ -0,0 +1,158 @@
...
+	if err != nil && !errors.Is(err, rgwadmin.ErrUserExists) {
+		klog.Error("failed to create user", err)
+		return nil, status.Error(codes.Internal, "User creation failed")
+	}
+
+	// TODO : Handle access policy in request, currently granting all perms to this user
+	policy, err := s.s3Client.GetBucketPolicy(bucketName)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok && aerr.Code() != "NoSuchBucketPolicy" {
+			return nil, status.Error(codes.Internal, "fetching policy failed")
```
Comment (leseb) [inline pkg/driver/provisioner.go:110]: failed to

File: pkg/driver/provisioner.go [hunk 39]
```diff
@@ -0,0 +1,158 @@
...
+		Actions(s3cli.AllowedActions...)
+	if policy == nil {
+		policy = s3cli.NewBucketPolicy(*statement)
+	} else {
+		policy = policy.ModifyBucketPolicy(*statement)
+	}
+	_, err = s.s3Client.PutBucketPolicy(bucketName, *policy)
+	if err != nil {
+		klog.Error("failed to set policy", err)
+		return nil, status.Error(codes.Internal, "puting policy failed")
```
Comment (leseb) [inline pkg/driver/provisioner.go:129]: failed to

File: pkg/driver/provisioner.go [hunk 40]
```diff
@@ -0,0 +1,158 @@
...
+	req *cosi.ProvisionerRevokeBucketAccessRequest) (*cosi.ProvisionerRevokeBucketAccessResponse, error) {
+
+	// TODO : instead of deleting user, revoke its permission and delete only if no more bucket attached to it
+	klog.Infof("Deleting user %q", req.GetAccountId())
+	if err := s.rgwAdminClient.RemoveUser(context.Background(), rgwadmin.User{
+		ID:          req.GetAccountId(),
+		DisplayName: req.GetAccountId(),
+	}); err != nil {
+		klog.Error("failed to Revoke Bucket Access")
+		return nil, status.Error(codes.Internal, "falied to Revoke Bucket Access")
```
Comment (leseb) [inline pkg/driver/provisioner.go:151]: failed to

**Review by leseb — CHANGES_REQUESTED** [review]
few small nits

**Review by leseb — APPROVED** [review]

**Comment by leseb** [comment]
@thotz I don't see any gh action yet but I see you added a `make`? Is it planned in another PR?

**Comment by thotz** [comment]
> @thotz I don't see any gh action yet but I see you added a `make`? Is it planned in another PR?

I just added `make` file building container image, no test cases or gh actions added till now. Will add as separate PR

**Comment by leseb** [comment]
> > @thotz I don't see any gh action yet but I see you added a `make`? Is it planned in another PR?
> 
> I just added `make` file building container image, no test cases or gh actions added till now. Will add as separate PR

Sounds, anything else before I merge this?

**Comment by thotz** [comment]
> > > @thotz I don't see any gh action yet but I see you added a `make`? Is it planned in another PR?
> > 
> > 
> > I just added `make` file building container image, no test cases or gh actions added till now. Will add as separate PR
> 
> Sounds good`, anything else before I merge this?

This can be base for the next set of work, it defines all the existing APIs provided by the COSI(still need to improve)

**Pull request merged**